core/server/master/src/test/java/alluxio/master/file/FileSystemMasterSyncMetadataMetricsTest.java

```diff
@@ -0,0 +1,506 @@
+    outputStream.close();
+    syncScheme =
+        new LockingScheme(new AlluxioURI("/dir0/file0"), InodeTree.LockPattern.READ, true);
+    inodeSyncStream =
+        new InodeSyncStream(syncScheme, mFileSystemMaster, RpcContext.NOOP,
+            DescendantType.NONE, ListStatusContext.defaults().getOptions().getCommonOptions(),
+            false, true, false, false);
+    inodeSyncStream.sync();
+    assertEquals(1 + 1, count.getCount());
+    assertEquals(1 + 1, success.getCount());
+    assertEquals(3 + 1, pathSuccess.getCount());
+    assertEquals(0 + 0, fail.getCount());
+    assertEquals(0 + 0, pathFail.getCount());
+    assertEquals(3 + 0, noChange.getCount());
+    assertEquals(0 + 0, skipped.getCount());
+
+    //sync "/dir0/file0" again
+    //path is existent in inodeTree and UFS and UFS is available ->success and on change
+    syncScheme =
+        new LockingScheme(new AlluxioURI("/dir0/file0"), InodeTree.LockPattern.READ, true);
+    inodeSyncStream =
+        new InodeSyncStream(syncScheme, mFileSystemMaster, RpcContext.NOOP,
+            DescendantType.NONE, ListStatusContext.defaults().getOptions().getCommonOptions(),
+            false, true, false, false);
+    inodeSyncStream.sync();
+    assertEquals(2 + 1, count.getCount());
+    assertEquals(2 + 1, success.getCount());
+    assertEquals(4 + 1, pathSuccess.getCount());
+    assertEquals(0 + 0, fail.getCount());
+    assertEquals(0 + 0, pathFail.getCount());
+    assertEquals(3 + 1, noChange.getCount());
+    assertEquals(0 + 0, skipped.getCount());
+
+    //!forceSync && !shouldSync ->skip
+    syncScheme =
+        new LockingScheme(new AlluxioURI("/dir0/file0"), InodeTree.LockPattern.READ, false);
+    inodeSyncStream =
+        new InodeSyncStream(syncScheme, mFileSystemMaster, RpcContext.NOOP,
+            DescendantType.NONE, ListStatusContext.defaults().getOptions().getCommonOptions(),
+            false,
+            false, false, false);
+    inodeSyncStream.sync();
+    assertEquals(3 + 1, count.getCount());
+    assertEquals(3 + 0, success.getCount());
+    assertEquals(5 + 0, pathSuccess.getCount());
+    assertEquals(0 + 0, fail.getCount());
+    assertEquals(0 + 0, pathFail.getCount());
+    assertEquals(4 + 0, noChange.getCount());
+    assertEquals(0 + 1, skipped.getCount());
+
+    //path is existent in inodeTree and UFS and UFS throws IOException -> success
+    mUfs.mIsThrowIOException = true;
+    syncScheme = new LockingScheme(new AlluxioURI("/dir0/file0"), InodeTree.LockPattern.READ, true);
+    inodeSyncStream = new InodeSyncStream(syncScheme, mFileSystemMaster, RpcContext.NOOP,
+        DescendantType.NONE, ListStatusContext.defaults().getOptions().getCommonOptions(),
+            false, true, false, false);
+    inodeSyncStream.sync();
+    assertFalse(mFileSystemMaster.getInodeTree().inodePathExists(new AlluxioURI("/dir0/file0")));
+    assertEquals(4 + 1, count.getCount());
+    assertEquals(3 + 1, success.getCount());
```

Comment:
Why does IOException counted as success? Can you point me to the code where the IOException is caught?

Reply:
First, the InodeSyncStream tries to get the fingerprint from UFS.
https://github.com/Alluxio/alluxio/blob/d2dc74d57c8b00e3f8b3082bdcfb58fae4d6c9ed/core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java#L617-L619
The BaseUnderFileSystem will catch the IOException and return an INVALID_UFS_FINGERPRINT.
https://github.com/Alluxio/alluxio/blob/4915e376aa8c2cfc3ee89fa985687be4023adbf6/core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java#L102-L119